Userland/Services/RequestServer/ConnectionFromClient.cpp

```diff
@@ -106,35 +106,40 @@ Messages::RequestServer::SetCertificateResponse ConnectionFromClient::set_certif
 }
 
 struct Job {
-    explicit Job(URL url)
-        : m_url(move(url))
-    {
-    }
-
     static Job& ensure(URL const& url)
     {
         if (auto it = s_jobs.find(url); it == s_jobs.end()) {
-            auto job = make<Job>(url);
+            auto job = adopt_own(*new Job(url));
             s_jobs.set(url, move(job));
         }
-        return *s_jobs.find(url)->value;
+        auto& job = *s_jobs.find(url)->value;
+        job.m_use_count++;
+        return job;
     }
 
     void start(Core::Socket& socket)
     {
         auto is_connected = socket.is_open();
         VERIFY(is_connected);
         ConnectionCache::request_did_finish(m_url, &socket);
-        s_jobs.remove(m_url);
+        if (--m_use_count <= 0)
+            s_jobs.remove(m_url);
     }
     void fail(Core::NetworkJob::Error error)
     {
         dbgln("Pre-connect to {} failed: {}", m_url, Core::to_string(error));
-        s_jobs.remove(m_url);
+        if (--m_use_count <= 0)
+            s_jobs.remove(m_url);
     }
-    URL m_url;
 
 private:
+    explicit Job(URL url)
+        : m_url(move(url))
+    {
+    }
+
+    URL m_url;
+    int m_use_count { 0 };
```

Comment:
hem. Perhaps making these guys RefCounted would do the trick in a more idiomatic way?

Reply:
we also have the magical `one_ref_left` function available to do the cache removal

Reply:
I've given that a try. It also works, though it does slightly change other users too. I placed a few logs around and it seems the jobs are still destroyed too.